rollup/rollup_sync_service/rollup_sync_service.go

```diff
@@ -0,0 +1,420 @@
+	if batchIndex == 0 {
+		return []*rawdb.ChunkBlockRange{{StartBlockNumber: 0, EndBlockNumber: 0}}, nil
+	}
+
+	tx, _, err := s.client.client.TransactionByHash(context.Background(), vLog.TxHash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get transaction, err: %w", err)
+	}
+
+	return s.decodeChunkBlockRanges(tx.Data())
+}
+
+// decodeChunkBlockRanges decodes chunks in a batch based on the commit batch transaction's calldata.
+func (s *RollupSyncService) decodeChunkBlockRanges(txData []byte) ([]*rawdb.ChunkBlockRange, error) {
+	const methodIDLength = 4
+	if len(txData) < methodIDLength {
+		return nil, fmt.Errorf("transaction data is too short")
+	}
+
+	method, err := s.scrollChainABI.MethodById(txData[:methodIDLength])
+	if err != nil {
+		return nil, fmt.Errorf("failed to get method by ID, ID: %v, err: %w", txData[:4], err)
+	}
+
+	values, err := method.Inputs.Unpack(txData[methodIDLength:])
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack transaction data using ABI: %v", err)
+	}
+
+	type commitBatchArgs struct {
+		Version                uint8
+		ParentBatchHeader      []byte
+		Chunks                 [][]byte
+		SkippedL1MessageBitmap []byte
+	}
+	var args commitBatchArgs
+	err = method.Inputs.Copy(&args, values)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode calldata into commitBatch args, err: %w", err)
+	}
+
+	if args.Version != batchHeaderVersion {
+		return nil, fmt.Errorf("unexpected batch version, expected: %d, got: %v", batchHeaderVersion, args.Version)
+	}
+
+	return DecodeChunkBlockRanges(args.Chunks)
+}
+
+// validateBatch verifies the consistency between the L1 contract and L2 node data.
+// The function will terminate the node and exit if any consistency check fails.
+// It returns the number of the end block, a finalized batch meta data, and an error if any.
+func validateBatch(event *L1FinalizeBatchEvent, parentBatchMeta *rawdb.FinalizedBatchMeta, chunks []*Chunk) (uint64, *rawdb.FinalizedBatchMeta, error) {
+	if len(chunks) == 0 {
+		return 0, nil, fmt.Errorf("invalid argument: length of chunks is 0")
+	}
+
+	startChunk := chunks[0]
+	if len(startChunk.Blocks) == 0 {
+		return 0, nil, fmt.Errorf("invalid argument: block count of start chunk is 0")
+	}
+	startBlock := startChunk.Blocks[0]
+
+	endChunk := chunks[len(chunks)-1]
+	if len(endChunk.Blocks) == 0 {
+		return 0, nil, fmt.Errorf("invalid argument: block count of end chunk is 0")
+	}
+	endBlock := endChunk.Blocks[len(endChunk.Blocks)-1]
+
+	localStateRoot := endBlock.Header.Root
+	if localStateRoot != event.StateRoot {
+		log.Error("State root mismatch", "batch index", event.BatchIndex.Uint64(), "start block", startBlock.Header.Number.Uint64(), "end block", endBlock.Header.Number.Uint64(), "parent batch hash", parentBatchMeta.BatchHash.Hex(), "l1 finalized state root", event.StateRoot.Hex(), "l2 state root", localStateRoot.Hex())
+		syscall.Kill(os.Getpid(), syscall.SIGTERM)
```

Comment:
This is what I said about batch validation. We just crash if local batch is different from what we get from L1